Browse through previous versions of a leaf

jzimdars opened the pull request:
Current version on the right, most recent previous version on the left. Hit the arrows to navigate earlier edits.

<img width="1711" alt="Screenshot 2024-05-21 at 3 28 51 PM" src="https://github.com/basecamp/workbook/assets/107673/376c3594-14ac-4eae-8e32-17eaef9ec0ae">

jzimdars committed:
Stub basic version viewer

kevinmcconnell committed:
Wire up version viewer

kevinmcconnell commented:
I've wired up the basics of this, and also remove the diffing code and simplified the edit tracking a little bit. I think we need some sort of placeholder for when the arrows aren't visible (or maybe show them as disabled?) so the layout in that heading doesn't more around. But it interaction is working:

https://github.com/basecamp/workbook/assets/1186763/d7cb5e1e-f299-46b1-b759-421b97d4a496

I have a couple of questions about the last parts of it:

- Should we do the same for picture pages, so you can see the previous pictures that you've replaced?
- Should editing titles be part of this? (And by extension, do we need a version view for title pages as well?)

There's also still the issue of an extra version being created when you make a new section. If we decide not to track title changes, that could solve that. Otherwise we'll need to suppress the versioning until the first body content edit.

jzimdars committed:
Clearly disabled buttons when you reach the edges of version history

jzimdars commented:
We talked about this on our planning call today but for the record:
1. I fixed the version navigation jumping in https://github.com/basecamp/workbook/pull/94/commits/bea1ee45970783cdacb6b0d3f66bbe65fa791d9f
2. We should show version for Picture and Section leaves if it's not a heavy lift. It's not as clearly useful in those cases but it is nice to have the same features available for all page types.
3. We decided to stop tracking titles, and are going to try to move the title to the Leaf model. Think of titles more like file names than actually part of the leaf content.

kevinmcconnell committed:
Preserve images when editing titles

Fixes two issues with editing leafables:

- For the title to be saved, it needs to be in the same form as the rest
  of the fields
- For Pictures, we were losing the image on edit because `dup`
  does not copy the attachments. Fixed by adding a new
  `dup_with_attachments` that will include them.
